client/core/core.go

```diff
@@ -1255,6 +1260,58 @@ func (c *Core) wallet(assetID uint32) (*xcWallet, bool) {
 	return w, found
 }
 
+// GetCachedPassword takes the mapping string passed to CacheAppPassword and
+// the key returned from it to lookup and decrypt the app password.
+func (c *Core) GetCachedPassword(mapping string, key []byte) ([]byte, error) {
+	_, pushes, err := encode.DecodeBlob(key)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding blob: %v", err)
+	}
+
+	key, deserializedCrypter := pushes[0], pushes[1]
+	crypter, err := encrypt.Deserialize(key, deserializedCrypter)
+	if err != nil {
+		return nil, fmt.Errorf("error deserializing crypter: %v", err)
+	}
+
+	c.encPassMtx.Lock()
+	encryptedPassword, ok := c.encryptedPasswords[mapping]
+	c.encPassMtx.Unlock()
+	if !ok {
+		return nil, fmt.Errorf("cached encrypted password not found for"+
+			" mapping: %v", mapping)
+	}
+
+	pw, err := crypter.Decrypt(encryptedPassword)
+	if err != nil {
+		return nil, fmt.Errorf("error decrypting password: %v", err)
+	}
+
+	return pw, nil
+}
+
+// CacheAppPassword encrypts the app password with a random key, caches it,
+// and returns the key. The mapping string argument is used to lookup the
+// cached password using the GetCachedPassword function.
+func (c *Core) CacheAppPassword(appPW []byte, mapping string) ([]byte, error) {
```

Comment:
Just a couple of notes on this approach. 

1. Even though I had originally suggested that `Core` be involved with the generation of these keys, based on our discussion since, I'm now thinking that we can just handle the entire thing in **webserver**. Is there any reason `Core` needs to be involved?
1. If `mapping` is a unique string associated with the generated key, then `mapping` should be generated by `Core`, perhaps stored in the key blob.

Reply:
`mapping` is  the `authToken` that's also in the cookies. I think you're right, this can just go in `webserver`.